File: Composite/cpp/Enemy.cpp
#include <bits/stdc++.h>
#include <list>
#include "Ball.cpp"

using namespace std;

class Enemy:public Entity{
    private:
        list<Entity*> myParts;
    public:
        Enemy(list<Entity*> newPars);
        virtual void logic();
        void add(Entity *newPart);
        void remove();
        Entity* getChild();
};

Enemy::Enemy(list<Entity*> newPars){
    myParts=newPars;
}

void Enemy::add(Entity *newPart){
    this->myParts.push_front(newPart);
}

void Enemy::remove(){
    this->myParts.pop_front();
}

Entity* Enemy::getChild(){
    return this->myParts.front();
}

void Enemy::logic(){
    cout << "Shoot:" << endl;
    this->getChild()->logic();
    this->remove();
}

File: Composite/Java/ClaseConcretas/Enemy.java
package ClaseConcretas;

import java.util.List;
import Entidad.*;

public class Enemy implements Entity {

    List<Entity> myParts;

    public Enemy( List<Entity> newParts){
        myParts=newParts;
    }

    public void logic(){
        System.out.println("Shoot:");
        getChild(0).logic();
        remove(0);
    }

    public void add(Entity newPart){
        myParts.add(newPart);
    }

    public void remove(int index){
        myParts.remove(index);
    }

    public Entity getChild(int index){
        return myParts.get(index);
    }
}
